### DISCUSSION TITLE
Flexible timeline_variable based on earlier blocks

### DISCUSSION THREAD

**Discussion by trig11** [jspsych/jsPsych#921] [Q&A] [answered]
Hi,

I am sorry if this might be a question with an obvious answer to it -- but I have been unable to find it in the documentation/examples.

The basic premise of what I am trying to do is that I want to have my subjects complete one practice block of the experiment, run some analyses on this practice data, and create a new timeline_variable based on some result properties of this practice block. For instance, let's say I want to calculate the median response time and use it as the stimulus duration for some of the trials in the next block (for which I want to create the new timeline_variable).

Obviously, if I just put something like var new_timeline = (...) into the html file, this loads right on the start-up of the experiment, where the necessary data is not yet available. I think I need some var of type: 'call-function' which is pushed to the timeline after the practice block -- but when I check this variable in the console I only get some default information that the variable is of the type 'call-function'. I think what I am asking is: how do I properly access the results of a call-function type variable and how can I use it as a timeline_variable in an upcoming block.

**Comment by jodeleeuw** [comment, marked as answer]
Hi @trig11 ,

It sounds like what you need here are [dynamic parameters](https://www.jspsych.org/overview/trial/#dynamic-parameters). This may avoid the need for a `call-function` trial, since you can just put the function directly in the parameter.

> how do I properly access the results of a call-function type variable

The return value of the function you call is stored in the trial data as `value`. You can access this with `jsPsych.data.get()` and manipulating the resulting [DataCollection](https://www.jspsych.org/core_library/jspsych-data/#datacollection). 

```js
var call_fn_trial = {
	type: 'call-function',
	func: function(){
		return 'Hello there.'
	},
	data: {unique_tag: 'stimulus_msg_trial'}
}

var next_trial = {
	type: 'html-keyboard-response',
	stimulus: function(){
		var msg = jsPsych.data.get().filter({unique_tag: 'stimulus_msg_trial'}).values()[0].value;
		// above looks complicated
		// it is getting all the data, .get()
		// filtering to any trial data that has the matching key/value pair, .filter()
		// extracting the raw data as an array, .values()
		// grabbing the first trial in that array, [0]
		// and getting the value property from that trial, .value
		return msg;
	}
}
```

> how can I use it as a timeline_variable in an upcoming block.

With the current version of jsPsych (`6.1.0`) timeline variables cannot change during the experiment. You can use dynamic parameters as timeline variables, but you can't modifying the timeline variables after the experiment has begun. We will probably find a way to make timeline variables adjustable mid-experiment in a future release to provide a simpler interface for this kind of design.

**Reply by trig11** [reply]
Thanks for this detailed reply, I found it very helpful.
In case anyone has a similar problem and finds this helpful, I have now solved my problem by using dynamic parameters for the trial duration and return a value from the results of my call_function trial:
```js
trial_duration: function(){
		var duration_label = jsPsych.timelineVariable('duration',true);
		var duration_array = jsPsych.data.get().filter({unique_tag: 'calc_tria_durationsl'}).values()[0].value;
		if (duration_label == 1) {
		     return duration_array[0];
		} else if (...)
		}
```
